Initial PSYKE support

TeranIvy opened the pull request:
This PR is for #139 to introduce initial support for kernel extraction: 
1) Marking nodes for extraction,
2) Adding comment to the PSy layer code to indicate the region marked for extraction,
3) ~~Generating a simple driver with extracted region (to be fully populated later),~~
4) Adding a simple user control via the optimisation script.

~~WIP: First two things are done but they need to be cleaned-up and the tests extended. There are basics for the latter for now. There are also several temporary files to be deleted when the work is finshed.~~

UPDATE: The third point is being addressed in issue #298.

arporter commented on src/psyclone/extractor.py:
```diff
@@ -0,0 +1,220 @@
+    settings for generating driver for the extracted code. '''
+
+    @staticmethod
+    def extract_kernel(schedule, kernel_name, root_node_position=None):
```
I'm sorry but (unless I've missed something) I'm really not convinced that this routine is needed. I think PSyKE is a tool for the HPC expert who will need to extract a kernel in the context of the PSy layer as a whole - i.e. including any OpenMP or OpenACC directives. The safest and most robust way to do this is to look at the Schedule and select the set of Nodes to extract. I think starting from the Kernel and going up is making life difficult as you need to allow for all sorts of things that could be an ancestor of the kernel (e.g. an OpenMP parallel directive or an OpenACC data region). This is highlighted by the fact that you've had to add a `depth` property to the Node - something that it should not have to care about since it cannot reliably be used to reason about things (since that requires us to predict all possible positions of a Node in a Schedule).

TeranIvy replied:
Removed the routine. As the `Extractor` class became empty I removed it as well. I will add it back in issue #298 as I will need it to wrap-up driver settings. For now I added the explanation in the module description with the issue number.

arporter commented on src/psyclone/generator.py:
```diff
@@ -309,7 +309,7 @@ def main(args):
     if args.script is not None and args.profile is not None:
         print("Error: use of automatic profiling in combination with an\n"
-              "optimisation script is not recommened since it may not work\n"
+              "optimisation script is not recommended since it may not work\n"
```
Thanks :-)

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -61,7 +61,7 @@ def colored(text, _):
         :param _: Fake argument, only required to match interface
                   provided by termcolor.colored
-        :return: The supplied text, unchanged
+        :returns: The supplied text, unchanged
```
Either `return` or `returns` is fine.

TeranIvy replied:
Sphinx user manual has only examples of `returns`.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -710,9 +711,9 @@ def unique_declns_by_intent(self, datatype):
-        :return: dictionary containing 'intent' keys holding the kernel
-                 argument intent and declarations of all kernel arguments
-                 for each type of intent
+        :returns: dictionary containing 'intent' keys holding the kernel
```
Please add line-continuation symbols.

TeranIvy replied:
Done.

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -5234,6 +5235,15 @@ def _setup(self, ktype, module_name, args, parent):
+        if self.name.lower().endswith("_code"):
+            self._base_name = self.name[:-5]
+        else:
+            # TODO: add a warning here when logging is added
```
Added

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -5439,26 +5449,22 @@ def local_vars(self):
 
+    @property
+    def base_name(self):
+        " Returns a base name for this kernel "
```
Done

TeranIvy commented on src/psyclone/extractor.py:
```diff
@@ -0,0 +1,220 @@
+            :param parent: the parent of this Node.
+            :type parent: a :py::class::`psyclone.psyGen.Node`.
+        '''
+        Node.__init__(self, children=children, parent=parent)
```
Done

TeranIvy commented on src/psyclone/extractor.py:
```diff
@@ -0,0 +1,220 @@
+        Node.__init__(self, children=children, parent=parent)
+
+    def __str__(self):
+        ''' Returns a string representation of the subtree starting at
```
Done

TeranIvy commented on src/psyclone/extractor.py:
```diff
@@ -0,0 +1,220 @@
+
+        :returns: the name of this Node, possibly with control codes for \
+                  colouring.
+        :rtype: string
```
Done

TeranIvy commented on src/psyclone/extractor.py:
```diff
@@ -0,0 +1,220 @@
+
+    @property
+    def dag_name(self):
+        ''' Return the base dag name for this ExtractNode '''
```
Done

TeranIvy commented on src/psyclone/extractor.py:
```diff
@@ -0,0 +1,220 @@
+        return "extract_" + str(self.position)
+
+    def view(self, indent=0):
+        ''' Print a text representation of this Extract schedule to stdout \
```
Removed the depth.

TeranIvy commented on src/psyclone/extractor.py:
```diff
@@ -0,0 +1,220 @@
+            entity.view(indent=indent + 1)
+
+    def gen_code(self, parent):
+        ''' Marks region for code extraction as children of the \
```
Amended.

TeranIvy commented on src/psyclone/extractor.py:
```diff
@@ -0,0 +1,220 @@
+
+    def gen_code(self, parent):
+        ''' Marks region for code extraction as children of the \
+        ExtractNode. For now it inserts comments at the position of the \
```
Amended.

TeranIvy commented on src/psyclone/extractor.py:
```diff
@@ -0,0 +1,220 @@
+        ''' Marks region for code extraction as children of the \
+        ExtractNode. For now it inserts comments at the position of the \
+        ExtractNode and after all children of the ExtractNode. These \
+        comments will later be replaced by calls to write out arguments \
```
Added issue number.

TeranIvy commented on examples/dynamo/eg12/extract_nodes.py:
```diff
@@ -0,0 +1,88 @@
+
+
+'''
+File containing a PSyclone transformation script for the Dynamo0p3 API to
```
Changed.

TeranIvy commented on examples/dynamo/eg12/compute_q_operator_kernel_mod.F90:
```diff
@@ -0,0 +1,185 @@
+  delta_z = dz
+end subroutine compute_q_operator_init
+
+!=============================================================================!
```
I removed the kernel from the example.

TeranIvy commented on examples/dynamo/eg12/dg_matrix_vector_kernel_mod.F90:
```diff
@@ -0,0 +1,139 @@
+                                 matrix,      &
+                                 ndf1, undf1, map1, &
+                                 ndf2, undf2, map2)
+ 
```
I edited this kernel slightly, however I did not remove Doxygen descriptions of input arguments as they are present in other LFRic examples as well. The kernel code is quite simple (no constructors or additional functions).

TeranIvy commented on examples/dynamo/eg12/colouring_and_omp.py:
```diff
@@ -0,0 +1,81 @@
+from psyclone.dynamo0p3 import DISCONTINUOUS_FUNCTION_SPACES
+
+
+def trans(psy):
```
Left it for now. I am looking into importing it from somewhere else.

arporter replied:
Although having duplication isn't great, I can't think of a better solution and I like the way you've built the extract transformation example on top of this transformation script. Therefore, I'm not going to worry about this.

TeranIvy commented on examples/dynamo/eg12/gw_mixed_schur_preconditioner_alg_mod.x90:
```diff
@@ -0,0 +1,463 @@
+!-----------------------------------------------------------------------------
```
Cut it down considerably so it requires only two kernels.

TeranIvy commented on examples/dynamo/eg12/matrix_vector_kernel_mod.F90:
```diff
@@ -0,0 +1,139 @@
+!-----------------------------------------------------------------------------
```
As above, I edited this kernel slightly, however I did not remove Doxygen descriptions of input arguments as they are present in other LFRic examples as well.

TeranIvy commented on examples/dynamo/README:
```diff
@@ -196,6 +196,46 @@ and manipulated:
     cd eg11/
     psyclone -s ./async_script.py helmholtz_solver_alg_mod.x90
 
+Example 12: Code extraction
```
Removed the `extract_single_kernel.py` script. Could not really think about the use case for `extract_kernel_multi_invokes.py` so I removed it, too. My idea here was to create multiple micro-benchmarks from one algorithm, however this would be better achieved by extracting nodes from different invokes than the same Kernel.
Renamed "optimisations" to "transformations" throughout and deleted a lot of code from the algorithm file. I also amended the `find_kernel.py` script to write out the name and schedule of Invokes containing the call to the specified Kernel.

TeranIvy commented on examples/dynamo/eg12/transpose_matrix_vector_kernel_mod.F90:
```diff
@@ -0,0 +1,139 @@
+!-----------------------------------------------------------------------------
```
Deleted the kernel altogether.

TeranIvy commented on doc/psyke.rst:
```diff
@@ -0,0 +1,423 @@
+      ! ExtractEnd
+
+The ``! CALL write_extract_arguments(argument_list)`` comment will be replaced
+by appropriate calls to write out arguments of extracted Node(s) or Kernel(s).
```
Added the reference.

TeranIvy commented on doc/psyke.rst:
```diff
@@ -0,0 +1,423 @@
+##################
+
+As noted in the :ref:`distributed_memory` section, support for distributed
+memory in PSyclone is currently limited the Dynamo0.3 API. Since the
```
Done.

TeranIvy commented on doc/psyke.rst:
```diff
@@ -0,0 +1,423 @@
+  ...
+
+As said above, extraction can be performed on optimised code. For example,
+the below example of transformation script first adds ``!$OMP PARALLEL DO``
```
Amended.

TeranIvy commented on doc/psyke.rst:
```diff
@@ -0,0 +1,423 @@
+++++++++
+
+Examples in ``examples/dynamo/eg12`` directory demonstrate how to
+apply code extraction by utilising PSyclone transformation scripts.
```
OK, I referred to the Examples chapter and deleted the entire "Utilities" section as it is now obsolete.

arporter replied:
Great.

TeranIvy commented on doc/psyke.rst:
```diff
@@ -0,0 +1,423 @@
+Shared memory and API-specific
+##############################
+
+As noted above, optimisations in the extracted code may fail if the extract
```
Deleted.

TeranIvy commented on doc/psyke.rst:
```diff
@@ -0,0 +1,423 @@
+++++++++++++
+
+Code extraction can be applied to unoptimised or optimised code. Hence, it
+is reasonable to introduce restrictions that prevent failures of optimising
```
Reworded.

TeranIvy commented on doc/psyke.rst:
```diff
@@ -0,0 +1,423 @@
+The ``! CALL write_extract_arguments(argument_list)`` comment will be replaced
+by appropriate calls to write out arguments of extracted Node(s) or Kernel(s).
+Both **ExtractRegionTrans** and **ExtractNode** are base classes that can be
+added to both Dynamo0.3 and GOcean1.0 APIs. However, generating argument list
```
Reworded, hopefully it is now clearer.

arporter replied:
Yes, thank you.

TeranIvy commented:
Hopefully all the remarks are addressed as agreed here and in the Skype call. Back to @arporter for second review.

arporter commented on doc/examples.rst:
```diff
@@ -142,3 +146,31 @@ Example of the use of transformations to introduce redundant computation,
+the full support for code extraction is being developed.
+
+This example also contains a Python helper script ``find_kernel.py``
+which displays useful information for Kernel extraction: names and
```
suggest "...displays the names and Schedules of Invokes containing call(s) to the specified Kernel:"

TeranIvy replied:
Done

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2825,3 +2824,216 @@ def validate(self, loop):
+            # Check that an ExtractNode is not already in the node list
+            # marked for extraction. Otherwise we would have an extract
+            # region within another extract region.
+            if node.walk(node.children, ExtractNode) or \
```
This check too will be handled by #292

TeranIvy replied:
It is so I removed the constraint and modified the relevant test.

arporter replied:
Mini-whoop!

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2825,3 +2824,216 @@ def validate(self, loop):
+                    "without its ancestor Directive is not allowed."
+                    .format(str(self.name)))
+
+            # Dynamo0.3 API constraint: Check that ExtractNode is not
```
This API-specific check should be in an API-specific sub-class.

TeranIvy replied:
Created sub-classes and modified tests accordingly.

arporter replied:
Great.

TeranIvy commented on src/psyclone/extractor.py:
```diff
@@ -0,0 +1,130 @@
+        :param int indent: depth of indent for output text.
+        '''
+        print(self.indent(indent) + self.coloured_text +
+              "[position='" + str(self.position) + "']")
```
Done.

TeranIvy commented on src/psyclone/extractor.py:
```diff
@@ -0,0 +1,130 @@
+        replaced by calls to write out arguments of extracted Node(s) or \
+        Kernel(s) in Issue #234.
+
+        :param parent: the parent of this Node.
```
Done.

TeranIvy commented on src/psyclone/psyGen.py:
```diff
@@ -1190,26 +1204,51 @@ def parent(self, my_parent):
+        a Node if found.
+        :param children: list of Nodes which are children of this Node
+        :type children: list of :py:class:`psyclone.psyGen.Node`
+        :param parent: the parent node of this Node
```
Oops, removed.

TeranIvy commented on src/psyclone/tests/extract_test.py:
```diff
@@ -0,0 +1,828 @@
+    assert ("Error in ExtractRegionTrans transformation: supplied nodes "
+            "are not children of the same Schedule/parent.") in str(excinfo)
+
+    if TEST_COMPILE:
```
I wasn't aware of that. Removed.

TeranIvy commented on src/psyclone/tests/extract_test.py:
```diff
@@ -0,0 +1,828 @@
+            "already contains another Extract region is not allowed.") \
+        in str(excinfo)
+
+    # Test GOcean1.0 API
```
Removed GOcean test.

TeranIvy commented on src/psyclone/tests/extract_test.py:
```diff
@@ -0,0 +1,828 @@
+    assert ("Extraction of a Kernel or a Built-in call without its "
+            "parent Loop is not allowed.") in str(excinfo)
+
+    # Test GOcean1.0 API for Kernel call error
```
Different entities are being exercised by the test - Built-in call (using an example from Dynamo 0.3 API) and Kernel Call (using and example from GOcean 1.0). If GOcean had built-ins the examples could be opposite. Therefore I left the test as it is.

arporter replied:
Ah yes. Thanks for explaining (I think I misunderstood the name of the test.)

TeranIvy commented on src/psyclone/tests/extract_test.py:
```diff
@@ -0,0 +1,828 @@
+            "allowed.") in str(excinfo)
+
+
+def test_loop_no_directive_gocean1p0():
```
I agree. OMP parallel loop transformations in both APIs are children of the same parent. Removed the test.

TeranIvy commented on src/psyclone/tests/extract_test.py:
```diff
@@ -0,0 +1,828 @@
+        "      ! ExtractEnd\n")
+    assert output in code
+
+    if TEST_COMPILE:
```
Removed.

TeranIvy commented on src/psyclone/tests/extract_test.py:
```diff
@@ -0,0 +1,828 @@
+        "      END DO \n")
+    assert output in code
+
+    if TEST_COMPILE:
```
Removed.

TeranIvy commented on src/psyclone/tests/extract_test.py:
```diff
@@ -0,0 +1,828 @@
+    assert extract_node[0].abs_position == abspos
+    assert extract_node[0].depth == dpth
+
+    # Test Dynamo0.3 API for extraction of a list of Nodes
```
I am not testing the APIs here, I am testing extraction of 1) a single node using and example from GOcean 1.0 API and 2) a list of nodes using an example from Dynamo 0.3 API. It could have been the opposite.

arporter replied:
Oh I see. Great.

TeranIvy commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2441,7 +2497,9 @@ def test_node_abstract_methods():
     from psyclone.psyGen import Node
     _, invoke = get_invoke("single_invoke.f90", "gocean1.0", idx=0)
     sched = invoke.schedule
+    sched.view()
```
Removed this and another occurrence of `sched.view()` (and the `print` below).

TeranIvy commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -1392,21 +1393,76 @@ def test_argument_backward_dependence(monkeypatch, annexed):
+        _, _ = child._find_position(child.root.children, -2)
+    assert "started from -2 instead of 0" in str(excinfo.value)
+
+    if TEST_COMPILE:
```
One of the development principles we have is to ultimately test all F90 examples for compilation and this is currently not done for "4.7_multikernel_invokes.f90", so I tested it here since I modified the test. I moved the compilation test to one of two tests in `dynamo0.3_transformations.test` that use this example.

arporter replied:
That's above and beyond - thanks!

TeranIvy commented on src/psyclone/tests/extract_test.py:
```diff
@@ -0,0 +1,828 @@
+    assert ("Error in ExtractRegionTrans: Distributed memory is not "
+            "supported.") in str(excinfo.value)
+
+    # Test GOcean1.0 API with distributed memory
```
Replaced the test another for `GlobalSum` with Dynamo0.3 API.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -2102,7 +2102,7 @@ def apply(self, outer):  # pylint: disable=arguments-differ
         :param outer: The node representing the outer loop.
         :type outer: :py:class:`psyclone.psyGen.Loop`
-        :return: A tuple consistent of the new schedule, and a Memento.
+        :returns: A tuple consistent of the new schedule, and a Memento.
```
Corrected.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -2825,3 +2824,216 @@ def validate(self, loop):
+        # constraint covers the presence of HaloExchange and GlobalSum
+        # classses as they are only generated when distributed memory
+        # is enabled.
+        if Config.get().distributed_memory:
```
Removed the constraint as it is now handled by `valid_node_types`. I amodified the former `test_distmem_error` test to just check for `HaloExchange` (now called `test_haloexchange_error`). I also added `test_globalsum_error` instead of the previous check for DistMeme in GOcean 1.0 API (comment above).

TeranIvy replied:
When I generated code with distributed memory enabled I remembered why I disabled it in the first place. The problem is that class `DynLoop` with distmem on generates infrastructure calls to set halos dirty/clean. I will use the nodes to generate driver layer and there is no simple way to remove the infrastructure calls. Therefore I enabled the restriction again and modified `test_distmem_errors` to check for `HaloExchange` and `GlobalSum` nodes as well.

arporter replied:
Oh, that's annoying. Never mind, we can always look at this in the future if it becomes a problem.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -2825,3 +2824,216 @@ def validate(self, loop):
+
+class ExtractRegionTrans(RegionTrans):
+    ''' Provides a transformation to extract code contained within one \
+    or more Nodes in a Schedule. For example:
```
Re-worded, thanks for suggestion.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -2825,3 +2824,216 @@ def validate(self, loop):
+    '''
+
+    def __str__(self):
+        return "Extract code within one or more nodes in an invoke."
```
Technically, this transformation creates a sub-tree with an `ExtractNode` as its root. A stand-alone program will be created by the `gen_driver` method of `ExtractNode` subclasses for different APIs. I re-worded the message accordingly.

TeranIvy commented on src/psyclone/tests/extract_test.py:
```diff
@@ -0,0 +1,828 @@
+            "Directive is not allowed.") in str(excinfo)
+
+
+def test_no_parent_accdirective():
```
They test different sub-classes of orphaned directives (OMP and ACC) so I left them as they are.

arporter replied:
On the face of it, yes, but in fact they both exercise the same check inside `ExtractRegionTrans._validate()`. Our approach is to write tests to acheive 100% coverage rather than to check every possible combination since that rapidly becomes impossible.

TeranIvy replied:
After modifying the constraint for orphaned directives in `ExtraxctRegionTrans` I realised your reasoning is correct. I deleted the preceding (OMP) test to so there is a better balance between Dynamo0.3 and GOcean1.0 code.

TeranIvy commented on src/psyclone/tests/extract_test.py:
```diff
@@ -0,0 +1,828 @@
+            "colour without its ancestor Loop over colours is not "
+            "allowed.") in str(excinfo)
+
+    # Now apply OMP Parallel DO Directive to the colour Loop and
```
Not really, as the basic constraint for extracting just the loop over cells without its ancestor loop over colours covers everything. It is just that colouring doesn't have much sense without OMP parallel Do directive. I modified the test accordingly and removed the duplicate check.

arporter replied:
Thanks.

TeranIvy commented on src/psyclone/transformations.py:
```diff
@@ -2825,3 +2824,216 @@ def validate(self, loop):
+        :raises TransformationError: if transformation is applied to an \
+                                     orphaned Directive without its parent \
+                                     Directive.
+        :raises TransformationError: if transformation is applied to a Loop \
```
Created sub-classes and modified tests accordingly.

TeranIvy commented on examples/dynamo/README:
```diff
@@ -196,6 +196,33 @@ and manipulated:
+::
+    cd eg12/
+    python find_kernel.py
+
```
Done.

TeranIvy commented:
Ready for 3rd review. I had to do some re-factoring due to changes introduced by #292 and sub-classing the `ExtractRegionTrans`. This, in turn, required modifying tests, examples and documentation. Checking for valid node list in #292 does not cover special cases here so I kept them where required. Also, I decided against lifting distributed memory restriction as this generates calls to LFRic infrastructure (see one of comments above) and it is not really clear to me how to remove them without further modification of Dynamo0.3 API.

Lastly, I am afraid this PR clashes with #285. The added algorithm example in `eg12` has invoke calls in function instead of subroutine. #285 does not support it so there is an issue #317 to extend it. It would be good to have a brief discussion how to take this forward.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2854,3 +2854,282 @@ def validate(self, loop):
+            # optimisations are applied. Note that we need to explicitly
+            # exclude the OMPParallelDoDirective as it inherits from both
+            # OMPDoDirective and OMPParallelDirective.
+            if isinstance(node, (OMPDoDirective, ACCLoopDirective)) and \
```
Upon reflection, I think that this check needs to be a bit more thorough. Currently this code assumes that any orphaned directives are immediate children of the Extract node. In practice, they may be further down than that (I'm thinking here of the NEMO API but it could be/become an issue for Dynamo and GOcean too). I think what we need to do is `node.walk()` down each sub-tree and check for directives. If the first directive we encounter in any of these walks is an orphaned directive then we raise an error.

arporter replied:
In fact, having just looked at your GOcean-specific sub-class of ExtractTrans I see there's a better way to do it - check on the types of any directives that are ancestors of the supplied list of nodes.

TeranIvy replied:
Done. Also deleted the test for the orphaned OMP directive and just left the one for the orphaned ACC directive from `extract_test.py. They are testing the same constraint as you commented.

arporter replied:
Nice - thanks.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -1199,26 +1213,49 @@ def parent(self, my_parent):
+        '''
         if self.parent is None:
-            return 0
+            return self.START_POSITION
```
This line is not covered.

TeranIvy replied:
Added the appropriate test in `test_node_position` (`psyGen_test.py`).

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -1199,26 +1213,49 @@ def parent(self, my_parent):
         if not found:
-            raise Exception("Error in search for my position in "
-                            "the tree")
+            raise InternalError("Error in search for Node position "
```
This line is not covered.

TeranIvy replied:
I had to modify the function `abs_position` in `class Node` to be able to reach that line. My idea was that search for a random Node that does not belong to the Schedule would trigger the exception. However, that did not happen because the condition `if self.root == self` did not check that the Node root is actually a Schedule instance (I guess that was implicitly assumed) and would return `START_POSITION`. I added the check that and then modified the `test_node_position` to test the constraint.

arporter commented on src/psyclone/tests/dynamo0p3_transformations_test.py:
```diff
@@ -949,7 +949,8 @@ def test_multi_kernel_single_omp_region():
 
-def test_multi_different_kernel_omp(monkeypatch, annexed):
+def test_multi_different_kernel_omp(
+        tmpdir, f90, f90flags, monkeypatch, annexed):
```
Sorry to be a pain but, since you've modified this test already, please could you modify it some more to use the `dist_mem` fixture?

TeranIvy replied:
Done. Very elegant solution using fixture, I must admit.

TeranIvy replied:
Also updated the compilation testing after merging changes from the latest master.

arporter replied:
Great, thanks.

TeranIvy commented:
I implemented `gen_c_code` method in `ExtractNode` as requested. I did not add a test for `NotImplementedError` as it does not seem obvious how I would do this now. As noted in the comments above, I added tests for uncovered lines and had to slightly modify one of conditions in `abs_position` to be able to reach the line at all. I modified the constraint for orphaned directives to check for ancestors and deleted one of the tests for it as requested. 
Also,
- the branch is updated with the latest master to pick up changes in test compilation,
- the function containing invoke calls in the algorithm example is modified to a subroutine so there are no clashes with the fparser2 branch.
I hope this covers all review remarks. Ready for the next round, @arporter.

TeranIvy commented:
**Update:** Added test for `gen_c_code` method of `ExtractNode` class. The branch is now ready for review.

arporter commented on src/psyclone/transformations.py:
```diff
@@ -2946,8 +2947,9 @@ def _validate(self, node_list):
+            if node.ancestor(OMPParallelDirective,
+                             excluding=[OMPParallelDoDirective]) or \
+                    node.ancestor(ACCParallelDirective):
                 raise TransformationError(
```
Strictly speaking, we're now forbidding the insertion of an Extraction region within a (thread) parallel region which seems eminently sensible to me. In fact, I think that means we shouldn't have the `excluding=[...]` bit as that also defines a parallel region. Please could you update and ensure that the error message reflects this.

TeranIvy replied:
You're right. Updated the constraint and the related test.

arporter replied:
Great, thanks.

arporter commented on doc/psyke.rst:
```diff
@@ -0,0 +1,287 @@
+
+* The entire Invoke (extraction applied to all Nodes).
+
+The basic mechanism of code extraction is applying the **ExtractRegionTrans**
```
"is through applying"

TeranIvy replied:
Done.

arporter commented on doc/psyke.rst:
```diff
@@ -0,0 +1,287 @@
+This group of restrictions is enforced irrespective of whether optimisations
+are used or not.
+
+* Extraction can be applied to a single Node in or a list of Nodes in a
```
s/Node in or/Node or/

TeranIvy replied:
Ooops. Corrected.

arporter commented on doc/psyke.rst:
```diff
@@ -0,0 +1,287 @@
+As noted in the :ref:`distributed_memory` section, support for distributed
+memory in PSyclone is currently limited to the Dynamo0.3 API. Since the
+implementation generates calls to LFRic infrastructure (e.g. runtime checks
+for status of field halos), code extraction when distributed memory is
```
"...extraction is not allowed when distributed..."

TeranIvy replied:
Rephrased, thanks.

arporter commented on doc/psyke.rst:
```diff
@@ -0,0 +1,287 @@
+Shared memory and API-specific
+##############################
+
+The restrictions below do not allow the **ExtractRegionTrans** to be
```
Suggest: "The **ExtractRegionTrans** transformation cannot be applied to:"

TeranIvy replied:
Definitely simpler. Done.

arporter commented on doc/psyke.rst:
```diff
@@ -0,0 +1,287 @@
+  without its parent Directive (e.g. ACC or OMP Parallel Directive),
+
+* A Loop over cells in a colour without its parent Loop over colours in
+  Dynamo0.3 API,
```
the Dynamo0.3 API.

TeranIvy replied:
Thanks, corrected. Give me a decade more in the UK and perhaps I'll understand articles :)

arporter replied:
I don't understand them. I just know when they should be there :-)

arporter commented on doc/psyke.rst:
```diff
@@ -0,0 +1,287 @@
+* A Loop over cells in a colour without its parent Loop over colours in
+  Dynamo0.3 API,
+
+* An inner Loop without its parent outer Loop in GOcean1.0 API.
```
the GOcean1.0 API.

TeranIvy replied:
Same as above.

TeranIvy commented:
Made the changes. Ready for another review.

arporter commented:
All requested changes have been made.
All tests pass/xfail (including with compilation enabled).
Documentation is up-to-date and builds OK. I shall take the liberty of moving the documentation on the two API-specific extract transformations to the relevant sections.
PR includes a relevant example.
PR does not affect LFRic functionality.

arporter commented:
This is just waiting on Travis now for one final check before I merge it to master.
